### PR TITLE
3.2 group page panchenko

### PR DESCRIPTION
**Pull request by antonpanchenk0**

### CODE REVIEW

File: front/src/containers/list-of-group/index.js [hunk 1]
```diff
@@ -1,65 +1,49 @@
...
     }
 }
 
-const mapStateToProps = ({ groupList: { currentGroup }, login: { userId } }) => ({ currentGroup, userId });
+const mapStateToProps = ({ groups: { groupList }, login: { userId } }) => ({ groupList, userId });
 
-const mapDispatchToProps = {
-    setCurrentGroup,
-};
+const mapDispatchToProps = (dispatch) => {
```
Comment (Alan18081) [inline front/src/containers/list-of-group/index.js:38]: Simplify it to plain object in format `{ onHandleGroup, onHandleLoadGroupData }`

File: front/src/components/group/index.js [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+                                    {
+                                        !isGroupEdited ?
+                                            <div>
+                                                {
+                                                    avatar?
+                                                    <CustomAvatar avatar={avatar.avatarLink} />
+                                                    :
+                                                    <CustomAvatar avatar={false} />
+                                                }
+                                                <GroupTitleP>{currentGroup.groupName}</GroupTitleP>
```
Comment (Alan18081) [inline front/src/components/group/index.js]: It should have name `GroupTitle`

File: front/src/components/group/index.js [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+import { checkUserStatus } from '../../common/functions';
+
+const { Title } = Typography;
+const { Content } = Layout;
+const { confirm } = Modal;
+
+class Group extends Component {
+    constructor(props){
+        super(props);
+        this.state = {
```
Comment (Alan18081) [inline front/src/components/group/index.js]: Please, move it to `redux`

File: front/src/common/redux/store.js [hunk 4]
```diff
@@ -7,10 +7,7 @@ import rootSaga from './root-saga';
 const sagaMiddleware = createSagaMiddleware();
 
 const store = createStore(
-  rootReducer,
-  window.__REDUX_DEVTOOLS_EXTENSION__ && window.__REDUX_DEVTOOLS_EXTENSION__(),
```
Comment (Alan18081) [inline front/src/common/redux/store.js:11]: Why did you remove devtools extension?

File: front/src/common/redux/groups/groups.action.js [hunk 5]
```diff
@@ -1,8 +1,41 @@
...
 
 export const setCurrentGroup = (data) => {
   return {
     type: SET_CURRENT_GROUP,
     data
   }
 };
+
+export const updateCurrentGroup = (data) => {
+  return {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.action.js]: Please, convert `return` to `= ({ ... })` and do it for every action in this file

File: front/src/common/functions/check-user-status.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+const checkUserStatus = () => {
+    const user = JSON.parse(localStorage.getItem('user'));
+    if(user.hasOwnProperty('isAdmin')){
```
Comment (Alan18081) [inline front/src/common/functions/check-user-status.js]: Don't use `user.hasOwnProperty` use just `user.isAdmin`

File: front/src/common/functions/check-user-status.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
...
+        return user.isAdmin ? 'superAdmin' : 'mentor';
```
Comment (Alan18081) [inline front/src/common/functions/check-user-status.js]: Please, define constants for `superAdmin`, `mentor` and `student`

File: front/src/common/redux/groups/groups.action.js [hunk 8]
```diff
@@ -1,8 +1,41 @@
...
+    type: UPDATE_CURRENT_GROUP,
+    data
+  }
+}
+
+export const startUpdatingGroupData = (data) => {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.action.js]: Why do you define arguments and don't use it inside the action

File: front/src/common/redux/groups/groups.action.js [hunk 9]
```diff
@@ -1,8 +1,41 @@
...
+  return {
+    type: START_LOAD_CURRENT_GROUP_DATA,
+  }
+}
+
+export const loadGroupsData = (data) => {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.action.js]: Why do you define arguments and don't use it inside the action

File: front/src/common/redux/groups/groups.action.js [hunk 10]
```diff
@@ -1,8 +1,41 @@
...
+  return {
+    type: LOAD_GROUPS_DATA,
+  }
+}
+
+export const changeUpdatingStatus = (data) => {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.action.js]: Why do you define arguments and don't use it inside the action

File: front/src/common/redux/groups/groups.action.js [hunk 11]
```diff
@@ -1,8 +1,41 @@
...
+  return {
+    type: CHANGE_UPDATING_STATUS_CURRENT_GROUP,
+  }
+}
+
+export const deleteStudentFromGroup = (data) => {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.action.js]: Please, give it a much more valid name, something like `id`, `data` is too broad

File: front/src/common/redux/groups/groups.action.js [hunk 12]
```diff
@@ -1,8 +1,41 @@
...
+  return {
+    type: DELETE_STUDENT_FROM_GROUP,
+    data,
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.action.js]: Use key `payload` for any action data

File: front/src/common/redux/groups/groups.reducer.js [hunk 13]
```diff
@@ -1,14 +1,39 @@
...
+        case types.COMPLETE_DELETE_STUDENT:
+            return { ...state, currentGroup: { ...action.data }, isDeleting: false };
+        case types.ERROR_UPDATE_CURRENT_GROUP:
+            return { ...state, groupsError: action.data, isUpdating: false, isGroupEdited: false };
+        case types.ERROR_DELETE_STUDENT:
+            return { ...state, groupsError: action.data };
+        case types.ERROR_LOADING_GROUPS_DATA:
+            return { ...state, groupsError: action.data };
+        default:
+            return { ...state };
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.reducer.js]: Just return `state` here, there is no need to copy it

File: front/src/common/redux/groups/groups.api.js [hunk 14]
```diff
@@ -0,0 +1,46 @@
+import axios from 'axios';
+
+async function updateGroup(request){
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.api.js]: Don't pass all action into API function, pass only data that it needs in an already prepared format. API function should be as simple as possible

Comment (Alan18081) [reply front/src/common/redux/groups/groups.api.js]: Don't mark function `async` if it doesn't return promise

File: front/src/common/redux/groups/groups.api.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+        data.avatar = request.data.avatar;
+    }
+    const headers = {
+        headers: {
+            'Content-Type': 'application/json',
+            'Authorization': `Bearer ${token}`,
+        }
+    };
+    return axios.put(url, data, headers)
+        .then(response => response)
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.api.js]: Please, return only data from API function, don't return the full response, saga should not know anything about HTTP requests

File: front/src/common/redux/groups/groups.api.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+async function getOneGroup(request) {
+    const token = localStorage.getItem('token');
+    const url = `${process.env.REACT_APP_GROUPS_LOAD_AND_UPDATE}${request.data.groupId}`;
+    const headers = {headers: {'Content-Type': 'application/json', 'Authorization': `Bearer ${token}`,}};
+    return axios.get(url, headers)
+        .then(response => response)
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.api.js]: Please, return only data from API function, don't return the full response, saga should not know anything about HTTP requests

File: front/src/common/redux/groups/groups.api.js [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+async function loadGroupsData(request) {
+    const token = localStorage.getItem('token');
+    const url = `${process.env.REACT_APP_GROUPS_LOAD_AND_UPDATE}`;
+    const headers = {headers: {'Content-Type': 'application/json', 'Authorization': `Bearer ${token}`,}};
+    return axios.get(url, headers)
+        .then(response => response)
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.api.js]: Please, return only data from API function, don't return the full response, saga should not know anything about HTTP requests

File: front/src/common/redux/groups/groups.api.js [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+async function deleteStudent(request) {
+    const { data: { studentId } } = request;
+    const token = localStorage.getItem('token');
+    const url = `${process.env.REACT_APP_DELETE_STUDENT_FROM_GROUP}${studentId}`;
+    const headers = {headers: {'Content-Type': 'application/json', 'Authorization': `Bearer ${token}`,}};
+    return axios.delete(url, headers)
+        .then(response => response);
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.api.js]: Please, return only data from API function, don't return the full response, saga should not know anything about HTTP requests

File: front/src/common/redux/groups/groups.saga.js [hunk 19]
```diff
@@ -0,0 +1,57 @@
+import { put, call, all, takeEvery } from 'redux-saga/effects';
+import { updateGroup, loadGroupsData, deleteStudent, getOneGroup } from './groups.api';
+import * as types from './types';
+
+export function* updateSaga(payload) {
+    try {
+        const response = yield call(updateGroup, payload);
+        yield put({type: types.SET_UPDATED_CURRENT_GROUP, data: response.data});
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.saga.js]: Please, use `payload` key instead of `data` and do it for all actions create as a part of this merge request

File: front/src/common/redux/groups/groups.saga.js [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+        const { data } = yield call(loadGroupsData, payload);
+        yield put({type: types.SET_GROUPS_DATA, data});
+    } catch (e) {
+        yield put({type: types.ERROR_LOADING_GROUPS_DATA, data: e});
+    }
+}
+
+export function* deleteStudentSaga(payload) {
+    try {
+        yield call(deleteStudent, payload);
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.saga.js]: It should be inside the `try {} catch` block

File: front/src/common/redux/groups/groups.saga.js [hunk 21]
```diff
@@ -0,0 +1,57 @@
...
+            yield put({type: types.ERROR_LOADING_GROUPS_DATA, data: e});
+        }
+
+    } catch (e) {
+        yield put({type: types.ERROR_DELETE_STUDENT, data: e});
+    }
+
+}
+
+function* watchUpdateGroup() {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.saga.js]: Get rid of these `watch` functions, move all `takeEvery` inside `all([ ... ])` in `groupSaga`

File: front/src/common/redux/groups/groups.selectors.js [hunk 22]
```diff
@@ -0,0 +1,6 @@
+export const updateGroupInGroupList = (groupList, newGroupData) =>{
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.selectors.js]: It is not a selector, the selector should only convert data from new format to another or get some parts of it, it should not modify it

Comment (Alan18081) [reply front/src/common/redux/groups/groups.selectors.js]: Data should be changed only inside reducer via actions

File: front/src/components/group/index.js [hunk 23]
```diff
@@ -0,0 +1,211 @@
...
+            role: checkUserStatus(),
+            fatalError: false,
+        };
+    }
+
+    componentDidCatch(error, errorInfo) {
```
Comment (Alan18081) [inline front/src/components/group/index.js]: When can fatal error happen?

File: front/src/components/group/index.js [hunk 24]
```diff
@@ -0,0 +1,211 @@
...
+        if(isDeleting && prevProps.isDeleting !== isDeleting) {
+            message.loading({ content: 'Loading...', key: 'delNotify' })
+        }
+        if(!isDeleting && prevProps.isDeleting) {
+            message.success({ content: 'Deleted!', key: 'delNotify', duration: 2});
+        }
+    }
+
+    handleErrorNotification = (error) => {
+        notification['error']({
```
Comment (Alan18081) [inline front/src/components/group/index.js]: Can you use `notification.error` here

File: front/src/components/group/index.js [hunk 25]
```diff
@@ -0,0 +1,211 @@
...
+            message: 'Something went wrong',
+            description: `${error}`,
```
Comment (Alan18081) [inline front/src/components/group/index.js]: We don't need template string here

File: front/src/components/group/index.js (hunk 2 above)
Comment (antonpanchenk0) [reply front/src/components/group/index.js]: done

File: front/src/components/group/index.js (hunk 3 above)
Comment (antonpanchenk0) [reply front/src/components/group/index.js]: done

File: front/src/common/redux/groups/groups.action.js (hunk 5 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.action.js]: done

File: front/src/common/functions/check-user-status.js (hunk 7 above)
Comment (antonpanchenk0) [reply front/src/common/functions/check-user-status.js]: done

File: front/src/common/functions/check-user-status.js (hunk 6 above)
Comment (antonpanchenk0) [reply front/src/common/functions/check-user-status.js]: done

File: front/src/common/redux/groups/groups.action.js (hunk 8 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.action.js]: done, deleted

File: front/src/common/redux/groups/groups.action.js (hunk 9 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.action.js]: done, deleted

File: front/src/common/redux/groups/groups.action.js (hunk 10 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.action.js]: done, deleted

File: front/src/common/redux/groups/groups.reducer.js (hunk 13 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.reducer.js]: done

File: front/src/common/redux/groups/groups.saga.js (hunk 21 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.saga.js]: done

File: front/src/components/group/index.js (hunk 24 above)
Comment (antonpanchenk0) [reply front/src/components/group/index.js]: done

File: front/src/components/group/index.js (hunk 25 above)
Comment (antonpanchenk0) [reply front/src/components/group/index.js]: done

File: front/src/common/redux/groups/groups.action.js (hunk 11 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.action.js]: Its realy data, because its obj {studentId, groupId}. We need groupId here to update group aftrer deleting student

File: front/src/common/redux/groups/groups.action.js (hunk 12 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.action.js]: done

File: front/src/common/redux/groups/groups.api.js (hunk 15 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.api.js]: done

File: front/src/common/redux/groups/groups.api.js (hunk 14 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.api.js]: done

File: front/src/common/redux/groups/groups.api.js (hunk 16 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.api.js]: done

File: front/src/common/redux/groups/groups.api.js (hunk 17 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.api.js]: done

File: front/src/common/redux/groups/groups.api.js (hunk 18 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.api.js]: done

File: front/src/containers/list-of-group/index.js (hunk 1 above)
Comment (antonpanchenk0) [reply front/src/containers/list-of-group/index.js:38]: done

File: front/src/common/redux/groups/groups.selectors.js (hunk 22 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.selectors.js]: done, selectors file is deleted

File: front/src/components/group/index.js (hunk 23 above)
Comment (antonpanchenk0) [reply front/src/components/group/index.js]: Just in case :relieved:

File: front/src/common/redux/groups/groups.saga.js (hunk 20 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.saga.js]: He is already in this block

File: front/src/common/redux/groups/groups.saga.js (hunk 19 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.saga.js]: done, fix in groups actions

File: front/src/components/group/index.js [hunk 26]
```diff
@@ -1,31 +1,26 @@
...
 
 class Group extends Component {
-    constructor(props){
-        super(props);
-        this.state = {
-            role: checkUserStatus(),
-            fatalError: false,
-        };
-    }
+    state = {
```
Comment (Alan18081) [inline front/src/components/group/index.js]: Please, remove this code, it is actually useless, in my opinion

Comment (antonpanchenk0) [reply front/src/components/group/index.js]: done

**Review by Alan18081 — APPROVED** [review]

File: front/src/common/redux/store.js (hunk 4 above)
Comment (antonpanchenk0) [reply front/src/common/redux/store.js:11]: fixed

File: front/src/common/redux/auth/auth.reducer.js [hunk 27]
```diff
@@ -1,8 +1,10 @@
 import * as types from './types';
+import { checkUserStatus } from '../../functions';
```
Comment (Alan18081) [inline front/src/common/redux/auth/auth.reducer.js]: Please, rename it to `getUserType`. It is more appropriate name

File: front/src/common/redux/groups/groups.saga.js [hunk 28]
```diff
@@ -0,0 +1,45 @@
...
+        const data = yield call(loadGroupsData);
+        yield put({type: types.SET_GROUPS_DATA, payload: data});
+    } catch (e) {
+        yield put({type: types.ERROR_LOADING_GROUPS_DATA, payload: e});
+    }
+}
+
+export function* deleteStudentSaga(action) {
+    try {
+        yield call(deleteStudent, action.payload);
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.saga.js:25]: Please, move it inside `try catch`

File: front/src/common/redux/groups/groups.saga.js [hunk 29]
```diff
@@ -0,0 +1,45 @@
+import { put, call, all, takeEvery } from 'redux-saga/effects';
+import { updateGroup, loadGroupsData, deleteStudent, getOneGroup } from './groups.api';
+import * as types from './types';
+
+export function* updateSaga(action) {
```
Comment (Alan18081) [inline front/src/common/redux/groups/groups.saga.js:5]: Please, remove all exports here except export for `groupSaga`

File: front/src/common/redux/auth/auth.reducer.js (hunk 27 above)
Comment (antonpanchenk0) [reply front/src/common/redux/auth/auth.reducer.js]: done

File: front/src/common/redux/groups/groups.saga.js (hunk 29 above)
Comment (antonpanchenk0) [reply front/src/common/redux/groups/groups.saga.js:5]: done

**Review by Alan18081 — APPROVED** [review]
